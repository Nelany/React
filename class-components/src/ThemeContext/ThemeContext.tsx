'use client';

import cookie from 'cookie';
import { createContext, ReactNode, useEffect, useState } from 'react';

export const ThemeContext = createContext({
  theme: 'light',
  toggleTheme: () => {},
});

interface ThemeProviderProps {
  children: ReactNode;
}

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [theme, setTheme] = useState('light');

  const getInitialTheme = () => {
    const document =
      typeof window !== 'undefined' ? window.document : { cookie: '' };
    const cookies = cookie.parse(document.cookie);

    return cookies.theme || 'light';
  };

  useEffect(() => {
    setTheme(getInitialTheme());
  }, []);

  const toggleTheme = () => {
    setTheme((currentTheme) => (currentTheme === 'light' ? 'dark' : 'light'));
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};
