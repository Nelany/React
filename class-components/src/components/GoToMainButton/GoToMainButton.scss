.go-to-main-button{
  margin: 40px 10px;
}