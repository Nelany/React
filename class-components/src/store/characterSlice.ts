import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from './store';
import { Character, CharacterResponse } from '../types/types';

interface CharLoadingProps {
  isCharLoading: boolean;
}

interface CharacterState {
  charactersResponse: CharacterResponse | null;
  selectedCharacters: Record<string, Character>;
  isCharLoading: CharLoadingProps;
}

const initialState: CharacterState = {
  charactersResponse: null,
  selectedCharacters: {},
  isCharLoading: {
    isCharLoading: false,
  },
};

const characterSlice = createSlice({
  name: 'characters',
  initialState,
  reducers: {
    setCharactersResponse: (state, action: PayloadAction<CharacterResponse | null>) => {
      state.charactersResponse = action.payload;
    },
    setSelectedCharacters: (state, action: PayloadAction<Character>) => {
      const character = action.payload;

      if (state.selectedCharacters[character.id]) {
        delete state.selectedCharacters[character.id];
      } else {
        state.selectedCharacters[character.id] = character;
      }
    },
    removeAllSelectedCharacters: (state) => {
      state.selectedCharacters = {};
    },
    setCharLoading: (state, action: PayloadAction<CharLoadingProps>) => {
      state.isCharLoading = action.payload;
    },
  },
});

export const {
  setCharactersResponse,
  setSelectedCharacters,
  removeAllSelectedCharacters,
  setCharLoading,
} = characterSlice.actions;

export const useIsCharLoading = () => {
  return useSelector((state: RootState) => state.characters.isCharLoading.isCharLoading);
};

export const useDispatchIsCharLoading = () => {
  const dispatch = useDispatch();

  return (isCharLoading: boolean) => {
    dispatch(setCharLoading({ isCharLoading }));
  };
};

export default characterSlice.reducer;