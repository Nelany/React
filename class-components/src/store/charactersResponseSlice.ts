import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { CharacterResponse } from '../types/types';

interface CharactersResponseState {
  data: CharacterResponse | null;
}

const initialState: CharactersResponseState = {
  data: null,
};

const charactersResponseSlice = createSlice({
  name: 'charactersResponse',
  initialState,
  reducers: {
    setCharactersResponse: (state, action: PayloadAction<CharacterResponse | null>) => {
      state.data = action.payload;
    },
  },
});

export const { setCharactersResponse } = charactersResponseSlice.actions;

export default charactersResponseSlice.reducer;