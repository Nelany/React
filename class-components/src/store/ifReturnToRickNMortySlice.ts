import { createSlice } from '@reduxjs/toolkit';

interface IfReturnToRickNMortyState {
  value: boolean;
}

const initialState: IfReturnToRickNMortyState = {
  value: false,
};

export const ifReturnToRickNMortySlice = createSlice({
  name: 'ifReturnToRickNMorty',
  initialState,
  reducers: {
    toggleIfReturnToRickNMorty: (state) => {
      state.value = !state.value;
    },
  },
});

export const { toggleIfReturnToRickNMorty } =
  ifReturnToRickNMortySlice.actions;

export default ifReturnToRickNMortySlice.reducer;