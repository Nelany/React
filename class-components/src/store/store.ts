import { configureStore } from '@reduxjs/toolkit';
import toastReducer from './toastSlice';
import { rtkApi } from '../api/rtkApi';
import { setupListeners } from '@reduxjs/toolkit/query';
import ifReturnToRickNMortyReducer from './ifReturnToRickNMortySlice';
import isCharLoading from './charactersLoadingSlice';
import charactersResponseReducer from './charactersResponseSlice';
import selectedCharactersReducer from './selectedCharactersSlice';


export const store = configureStore({
  reducer: {
    toast: toastReducer,
    [rtkApi.reducerPath]: rtkApi.reducer,
    ifReturnToRickNMorty: ifReturnToRickNMortyReducer,
    isCharLoading: isCharLoading,
    charactersResponse: charactersResponseReducer,
    selectedCharacters: selectedCharactersReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(rtkApi.middleware),
  devTools: true,
});

setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
