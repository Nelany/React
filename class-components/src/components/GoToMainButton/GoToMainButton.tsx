import { useNavigate } from 'react-router-dom';
import './GoToMainButton.scss';

export const GoToMainButton = () => {
  const navigate = useNavigate();

  const navigateToMain = () => {
    localStorage.setItem('searchQuery', '');
    navigate('/');
    window.location.reload();
  };

  return (
    <button className="go-to-main-button" onClick={navigateToMain}>
      Return to Rick and Morty!
    </button>
  );
};
