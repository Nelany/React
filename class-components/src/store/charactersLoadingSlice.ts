import { createSlice } from '@reduxjs/toolkit';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from './store';
import type { PayloadAction } from '@reduxjs/toolkit';

type CharLoadingProps = { isCharLoading: boolean };

export interface CharLoadingState {
  value: CharLoadingProps;
}

const initialState: CharLoadingState = {
  value: {
    isCharLoading: false,
  },
};

export const charactersLoadingSlice = createSlice({
  name: 'isCharactersLoading',
  initialState,
  reducers: {
    setCharLoading: (state, action: PayloadAction<CharLoadingProps>) => {
      state.value = action.payload;
    },
  },
});

export const { setCharLoading } = charactersLoadingSlice.actions;

export const useIsCharLoading = () => {
  return useSelector(
    (state: RootState) => state.isCharLoading.value.isCharLoading
  );
};

export const useDispatchIsCharLoading = () => {
  const dispatch = useDispatch();

  return (isCharLoading: boolean) => {
    dispatch(setCharLoading({ isCharLoading }));
  };
};

export default charactersLoadingSlice.reducer;
