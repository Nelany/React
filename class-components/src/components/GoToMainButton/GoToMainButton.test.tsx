import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { vi } from 'vitest';
import { store } from '../../store/store';
import { ThemeProvider } from '../../ThemeContext/ThemeContext';
import { GoToMainButton } from './GoToMainButton';

vi.mock('@remix-run/react', async () => {
  const actual = await vi.importActual('@remix-run/react');

  return {
    ...actual,
    useNavigate: () => vi.fn(),
  };
});

describe('GoToMainButton component', () => {
  test('matches snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <ThemeProvider>
          <GoToMainButton />
        </ThemeProvider>
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
