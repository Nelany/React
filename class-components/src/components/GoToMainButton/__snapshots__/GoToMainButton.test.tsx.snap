// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GoToMainButton component > matches snapshot 1`] = `
<button
  class="go-to-main-button"
>
  Return to Rick and Morty!
</button>
`;
